### PR TITLE
Drop beacon_chain pubkey to index map cache

### PR DESCRIPTION
**Pull request by dapplion**
## Issue Addressed

With current spec, the pubkey to index map could be a global cache. Right now this cache is duplicated in:
- once in BeaconChain struct
- one copy per BeaconState instance

This PR drops the copy in the BeaconChain struct, making its consumers use _some_ state instead.

### CODE REVIEW

File: beacon_node/store/src/hot_cold_store.rs [hunk 1]
```diff
@@ -1142,6 +1144,14 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
         }
     }
 
+    /// Rebase state pubkey_cache on some existing state to prevent re-populating the cache on the
+    /// first process_deposit call.
+    fn rebase_caches(&self, state: &mut BeaconState<E>) {
+        if let Ok((_, first_state)) = self.state_cache.lock().iter().exactly_one() {
+            state.rebase_caches_on(first_state);
+        }
+    }
```
Comment (dapplion) [inline beacon_node/store/src/hot_cold_store.rs:1153]: @michaelsproul is it possible to access the head state here? 
(2) Those the state_cache always include the head state? 
(3) Is the state_cache guaranteed to never be empty?

File: beacon_node/operation_pool/src/lib.rs [hunk 2]
```diff
@@ -1529,7 +1529,8 @@ mod release_tests {
         let (harness, _) = sync_contribution_test_state::<MainnetEthSpec>(1).await;
 
         let op_pool = OperationPool::<MainnetEthSpec>::new();
-        let state = harness.get_current_state();
+        let mut state = harness.get_current_state();
+        state.update_pubkey_cache().unwrap();
```
Comment (dapplion) [inline beacon_node/operation_pool/src/lib.rs:1533]: @michaelsproul could not figure out where the caches are lost. The genesis state in the harness builds the caches but they are lost along the way. Maybe while advancing the blocks?

File: beacon_node/store/src/hot_cold_store.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/store/src/hot_cold_store.rs:1153]: Oh, on `unstable` that `state_cache` is just for historic states. It would be better named `historic_state_cache`.

The `state_cache` on `tree-states` _can_ guarantee that the **finalized** state is always present. See:

https://github.com/sigp/lighthouse/blob/928915c71885a2439f8fa3b3a22ef6b39e978339/beacon_node/store/src/state_cache.rs#L35

(the `Option` is just for first-time initialization)

In summary:

1) No, but you could get the `finalized` state on `tree-states`. I think that would work?
2) Definitely not on unstable (head is not historic), and _probably yes_ (but not guaranteed) on `tree-states`.
3) On unstable, no; on tree-states, yes.

File: beacon_node/operation_pool/src/lib.rs (hunk 2 above)
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:1533]: Not sure, can dig into this if it's still a problem later

File: beacon_node/store/src/hot_cold_store.rs (hunk 1 above)
Comment (dapplion) [reply beacon_node/store/src/hot_cold_store.rs:1153]: Might be best to postpone this PR for post tree-states then
